src/cli.rs

```diff
@@ -606,77 +579,66 @@ fn open_channel(
 	}
 }
 
-fn send_payment(
-	payee: PublicKey, amt_msat: u64, final_cltv: u32, payment_hash: PaymentHash,
-	payment_secret: Option<PaymentSecret>, payee_features: Option<InvoiceFeatures>,
-	route_hints: Vec<&RouteHint>,
-	router: Arc<NetGraphMsgHandler<Arc<dyn chain::Access + Send + Sync>, Arc<FilesystemLogger>>>,
-	channel_manager: Arc<ChannelManager>, payment_storage: PaymentInfoStorage,
-	logger: Arc<FilesystemLogger>,
+fn send_payment<E: EventHandler>(
+	invoice_payer: &InvoicePayer<E>, invoice: &Invoice, payment_storage: PaymentInfoStorage,
 ) {
-	let network_graph = &router.network_graph;
-	let first_hops = channel_manager.list_usable_channels();
-	let payer_pubkey = channel_manager.get_our_node_id();
-
-	let route = router::get_route(
-		&payer_pubkey,
-		&network_graph,
-		&payee,
-		payee_features,
-		Some(&first_hops.iter().collect::<Vec<_>>()),
-		&route_hints,
-		amt_msat,
-		final_cltv,
-		logger,
-		&Scorer::default(),
-	);
-	if let Err(e) = route {
-		println!("ERROR: failed to find route: {}", e.err);
-		return;
-	}
-	let status = match channel_manager.send_payment(&route.unwrap(), payment_hash, &payment_secret)
-	{
+	let status = match invoice_payer.pay_invoice(invoice) {
 		Ok(_payment_id) => {
-			println!("EVENT: initiated sending {} msats to {}", amt_msat, payee);
+			let payee_pubkey = invoice.recover_payee_pub_key();
+			let amt_msat = invoice.amount_milli_satoshis().unwrap();
+			println!("EVENT: initiated sending {} msats to {}", amt_msat, payee_pubkey);
+			print!("> ");
 			HTLCStatus::Pending
 		}
-		Err(e) => {
+		Err(PaymentError::Invoice(e)) => {
+			println!("ERROR: invalid invoice: {}", e);
+			print!("> ");
+			return;
+		}
+		Err(PaymentError::Routing(e)) => {
+			println!("ERROR: failed to find route: {}", e.err);
+			print!("> ");
+			return;
+		}
+		Err(PaymentError::Sending(e)) => {
 			println!("ERROR: failed to send payment: {:?}", e);
+			print!("> ");
 			HTLCStatus::Failed
```

Comment:
Hm yeah I don’t think either would make sense to put in PaymentStorage. Don’t recall my exact reasoning tho.

Reply:
Yea, I think don't bother to store them? I'm not even sure if we need to store outbound payments anyway at this point?

Reply:
> I'm not even sure if we need to store outbound payments anyway at this point?

`list_payments` currently allows you to see whether an outbound payment is pending vs completed/failed, due to storing them here. Could be exposed from `ChannelManager` instead tho

Reply:
Ah, yea, I've thought we need to expose that from ChannelManager anyway, lets maybe drop it from the sample for now and then expose it upstream.

Reply:
Shouldn't we expose them upstream before dropping it (i.e., `outbound_payments`) from the sample so that `list_payments` doesn't break? Or did I misunderstand what should be dropped?

Reply:
I mean we don't have any API consistency guarantees in the sample, might as well drop something that doesn't quite work rather than tell users do to something that they won't need to do in a few versions IMO :)

Reply:
Opened https://github.com/rust-bitcoin/rust-lightning/issues/1157

Reply:
Ah, yeah and right now neither inbound nor outbound payments are persisted in the sample. They're just used in `list_payments`. I can drop them entirely. Just want to note that it seems independent of the PR. Happy to do it here in a follow-up.

Reply:
Fair, lets do it in a followup and land this for now.

Reply:
Hmm... so won't `ChannelManager` eventually remove the payments after they've been fulfilled? The sample would still needs to keep track of these, right? The idea of `list_payments` is basically a transaction log, IIUC.

Reply:
> Hmm... so won't `ChannelManager` eventually remove the payments after they've been fulfilled? The sample would still needs to keep track of these, right? The idea of `list_payments` is basically a transaction log, IIUC.

I second this, though payments are currently not persisted to disk in the sample. So IMO inbound/outbound tracking shouldn’t be removed